toolchain/parse/handle_decl_scope_loop.cpp

```diff
@@ -61,64 +71,113 @@ auto HandleDeclScopeLoop(Context& context) -> void {
+                                   context.Consume());
+    // Reuse state here to retain its `subtree_start`
+    state.state = next_state;
+    context.PushState(state);
+  };
+
+  bool saw_modifier = false;
+  while (true) {
+    switch (context.PositionKind()) {
+      // If we see a declaration modifier keyword token, add it as a leaf node
+      // and repeat with the next token.
+      case Lex::TokenKind::Abstract:
+      case Lex::TokenKind::Base:
+      case Lex::TokenKind::Default:
+      case Lex::TokenKind::Final:
+      case Lex::TokenKind::Override:
+      case Lex::TokenKind::Private:
+      case Lex::TokenKind::Protected:
+      case Lex::TokenKind::Virtual: {
         auto modifier_token = context.Consume();
-        auto class_token = context.Consume();
-        context.AddLeafNode(NodeKind::ClassIntroducer, class_token);
-        context.AddLeafNode(position_kind == Lex::TokenKind::Abstract
-                                ? NodeKind::AbstractModifier
-                                : NodeKind::BaseModifier,
-                            modifier_token);
+        context.AddLeafNode(NodeKind::DeclModifierKeyword, modifier_token);
+        saw_modifier = true;
+        break;
+      }
+
+      // If we see a declaration introducer keyword token, replace the
+      // placeholder node and switch to a state to parse the rest of the
+      // declaration. We don't allow namespace or empty declarations here since
+      // they can't have modifiers and don't use bracketing parse nodes that
+      // would allow a variable number of modifier nodes.
+      case Lex::TokenKind::Class: {
+        introducer(NodeKind::ClassIntroducer,
+                   State::TypeAfterIntroducerAsClass);
+        return;
+      }
+      case Lex::TokenKind::Constraint: {
+        introducer(NodeKind::NamedConstraintIntroducer,
+                   State::TypeAfterIntroducerAsNamedConstraint);
+        return;
+      }
+      case Lex::TokenKind::Fn: {
+        introducer(NodeKind::FunctionIntroducer, State::FunctionIntroducer);
+        return;
+      }
+      case Lex::TokenKind::Interface: {
+        introducer(NodeKind::InterfaceIntroducer,
+                   State::TypeAfterIntroducerAsInterface);
+        return;
+      }
+      case Lex::TokenKind::Var: {
+        introducer(NodeKind::VariableIntroducer, State::VarAsDecl);
+        return;
+      }
+      case Lex::TokenKind::Let: {
+        introducer(NodeKind::LetIntroducer, State::Let);
+        return;
+      }
+
+        // We don't allow namespace or empty declarations after a modifier since
+        // they can't have modifiers and don't use bracketing parse nodes that
+        // would allow a variable number of modifier nodes.
+
+      case Lex::TokenKind::Namespace: {
+        if (saw_modifier) {
+          HandleUnrecognizedDecl(context, state.subtree_start);
+        } else {
+          introducer(NodeKind::NamespaceStart, State::Namespace);
+        }
```

Comment:
FWIW I expect we'll want modifiers for namespaces, e.g.:
https://clang.llvm.org/docs/AttributeReference.html#abi-tag